book: add a QUIC page (#3380)
firedancer-io/firedancer: book/guide/quic.md

# QUIC in Firedancer

> Certain Solana network protocols are layered on top of QUIC ([RFC 9000]),

Reviewer: That's supposed to be an anchor link

the RFC documents.  These are precisely specified below.

## Performance

fd_quic is entirely single threaded.  Thus, all figures are per core.
fd_quic can be linearly scaled by L4 load balancing ingress to multiple
cores (receive side scaling).

As of 2024-Nov, fd_quic is optimized for use as a Solana TPU server.
It supports up to 1 million concurrent connections.

## QUIC-TLS Handshake

### TLS message fragmentation

fd_quic will fail to establish a connection to peers that fragment a TLS
handshake message (e.g. a ClientHello) across two or more QUIC CRYPTO
frames.  In other words, each TLS message must begin and end within one
QUIC CRYPTO frame, and thus also be entirely contained within the same
packet.  A CRYPTO frame may contain multiple handshake messages
concatenated.

The rationale is that virtually any TLS handshake exchange on the Solana
network only amounts to a few hundred bytes.

### Certificates

fd_quic is primarily designed for use in peer-to-peer networks which do not
have a concept of certificate authorities.  In Solana's peer-to-peer
protocols, the only useful piece of information in a certificate is the

> supposed public key of the peer.

Author: @two-heart I wanted to translate "der angebliche Public Key des Peers". The certificate says which public key you claim to hold and separately the CertificateVerify actually verifies it

fd_quic/fd_tls understand two certificate types to wrap the Ed25519 public
key: RFC 7250 Raw Public Keys (RPK), and fake X.509 certificates.

RPKs are the preferred certificate type.

> Because support for RPK is rare the wild, fd_quic further supports fake

Reviewer 2: ```suggestion
Because support for RPK is rare in the wild, fd_quic further supports fake
```

> Because support for RPKs are rare, fd_quic further supports fake X.509

Author: ```suggestion
Because support for RPK is rare the wild, fd_quic further supports fake X.509
```

000000b0: 0506 032b 6570 0341 00ff ffff ffff ffff  ...+ep.A........
000000c0: ffff ffff ffff ffff ffff ffff ffff ffff  ................
000000d0: ffff ffff ffff ffff ffff ffff ffff ffff  ................
000000e0: ffff ffff ffff ffff ffff ffff ffff ffff  ................
000000f0: ffff ffff ffff ffff ff                   .........
```

</details>

### Raw Public Keys

Raw Public Keys are an alternative certificate type specified in
[RFC 7250](https://datatracker.ietf.org/doc/html/rfc7250).

fd_quic/fd_tls support Ed25519 raw public keys.

<details>
<summary>Raw Public Key ASN.1 dump</summary>

```
    0:d=0  hl=2 l=  42 cons: SEQUENCE
    2:d=1  hl=2 l=   5 cons: SEQUENCE
    4:d=2  hl=2 l=   3 prim: OBJECT            :Ed25519
    9:d=1  hl=2 l=  33 prim: BIT STRING
```

</details>

<details>
<summary>Raw Public Key Hexdump</summary>

To construct a raw public key, paste the Ed25519 public key at offset 0x0c.

```
00000000: 302a 3005 0603 2b65 7003 2100 ffff ffff  0*0...+ep.!.....
00000010: ffff ffff ffff ffff ffff ffff ffff ffff  ................
00000020: ffff ffff ffff ffff ffff ffff            ............
```

</details>

### Next Protocol Selection

> fd_quic/fd_tls uses the ALPN TLS extension (RFC 7301) to verify that

Reviewer 2: Sometimes "fd_quic/fd_tls" is used as singular, other times as plural. Either one is fine, but keep it consistent.

does not include network and processing delays.

The delay logic is as follows:
- ACKs for handshake traffic and exceptional events are always sent without
  artificial delay.
- Each connection has an ACK delay buffer.
- ACKs for other traffic (e.g. application traffic / STREAM frames) are
  merged into the ACK delay buffer.
- The buffer is flushed if the oldest ACK is older than `max_ack_delay`.
- The buffer is flushed if more than `ack_threshold` bytes of stream data
  sent by the peer are awaiting ACKs.

Note that the ACK delay buffer is optimized towards ACKs over a gapless
sequence of packet numbers.  If the peer frequently skips packet numbers,
ACKs will get dropped before being sent.

(More details on the algorithm can be found in fd_quic_ack_tx)

## Encryption keys

fd_quic supports exporting of TLS master secrets in the SSLKEYLOGFILE
format.  fd_quic is further able to generate pcapng files that contain
both QUIC packets and TLS master secrets, allowing for convenient packet
analysis in Wireshark.

TLS secret logging is disabled by default.  It is available via the
`--pcap` flag for most fd_quic unit tests.

## Unsupported Features

> As of 2024-Nov, features that fd_quic does not support include:

Reviewer 2: Preferably use "November 2024" or alternatively "2024-11"
